client/txn.go

```diff
@@ -384,47 +378,40 @@ func (txn *Txn) exec(retryable func(txn *Txn) error) error {
 // always commit or clean-up explicitly even when that may not be
 // required (or even erroneous).
 func (txn *Txn) send(calls ...proto.Call) error {
+	if txn.Proto.Status != proto.PENDING {
+		return util.Errorf("attempting to use %s transaction", txn.Proto.Status)
+	}
+
 	lastIndex := len(calls) - 1
 	if lastIndex < 0 {
 		return nil
 	}
 
+	lastReq := calls[lastIndex].Args
 	// haveTxnWrite tracks intention to write. This is in contrast to
 	// txn.Proto.Writing, which is set by the coordinator when the first
 	// intent has been created, and which lives for the life of the
 	// transaction.
-	var haveTxnWrite bool
-	var haveEndTxn bool
+	haveTxnWrite := proto.IsTransactionWrite(lastReq)
 
-	for i, call := range calls {
+	for _, call := range calls[:lastIndex] {
```

Comment:
> The write status of the last call does matter, 

Yes, but that status is guaranteed to be reflected in `txn.Proto.Writing` before any `EndTransaction` call will be generated. So it matters, but it doesn't matter that it be included in `haveTxnWrite`. I'm just saying that either A) this loop should look at every request in the batch instead of dropping the last one (so that `haveTxnWrite` *means* `haveTxnWrite`) or B) we need a comment explaining why it's ok to ignore the last request (wordsmith however you like).

> I don't understand why we'd want to push any logic out to those places

Yeah, pushing the logic out isn't a great idea. But per my previous reply, once we have end-to-end batching, this logic could probably just be removed.